src/main/java/org/onedatashare/server/service/BoxService.java

```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Override
+    public Mono<Stat> mkdir(String cookie, UserAction userAction) {
+        return getBoxResourceUserActionUri(cookie, userAction)
+                .flatMap(BoxResource::mkdir)
+                .flatMap(BoxResource::stat);
+    }
+
+    @Override
+    public Mono<BoxResource> delete(String cookie, UserAction userAction) {
+        return getBoxResourceUserActionUri(cookie, userAction)
+                .flatMap(BoxResource::delete);
+    }
+
+    @Override
+    public Mono<Job> submit(String cookie, UserAction userAction) {
+        return null;
+    }
+
+    @Override
+    public Mono<String> download(String cookie, UserAction userAction) {
+        return getBoxResourceUserActionUri(cookie, userAction)
+                .flatMap(BoxResource::download);
+    }
+
+    public Mono<BoxResource> getBoxResourceUserActionUri(String cookie, UserAction userAction) {
+        final String path = pathFromUri(userAction.uri);
+        String id = userAction.id;
+        ArrayList<IdMap> idMap = userAction.map;
+        return userService.getLoggedInUser(cookie)
+                .map(User::getCredentials)
+                .map(uuidCredentialMap -> uuidCredentialMap.get(UUID.fromString(userAction.credential.getUuid())))
+                .map(credential -> new BoxSession(URI.create(userAction.uri), credential))
+                .flatMap(BoxSession::initialize)
+                .flatMap(boxSession -> boxSession.select(path, id, idMap)
+                        .onErrorResume(throwable -> throwable instanceof TokenExpiredException, throwable ->
+                                Mono.just(userService.updateCredential(cookie,((TokenExpiredException)throwable).cred))
+                                        .map(credential -> new BoxSession(URI.create(userAction.uri), credential))
+                                        .flatMap(BoxSession::initialize)
+                                        .flatMap(driveSession -> driveSession.select(path,id, idMap)))
+                );
+    }
+
+    public Mono<BoxResource> getBoxResourceUserActionResource(String cookie, UserActionResource userActionResource) {
+        final String path = pathFromUri(userActionResource.uri);
+        String id = userActionResource.id;
+        ArrayList<IdMap> idMap = userActionResource.map;
+
+        return userService.getLoggedInUser(cookie)
+                .map(User::getCredentials)
+                .map(uuidCredentialMap ->
+                        uuidCredentialMap.get(UUID.fromString(userActionResource.credential.getUuid())))
+                .map(credential -> new BoxSession(URI.create(userActionResource.uri), credential))
+                .flatMap(BoxSession::initialize)
+                .flatMap(boxSession -> boxSession.select(path, id, idMap));
+    }
+
+
+    public String pathFromUri(String uri) {
+        String path = "";
+        if(uri.contains("box://")){
+            path = uri.split("box://")[1];
```

Comment:
Not a good way of doing this. Please see how this operation has been handled in other classes in latest code.

Reply:
Updated to the better way

Reply:
Looks good now!